### ISSUE TITLE
ILLinkTrimAssembly condition applies to NetCoreAppCurrent only

### ISSUE THREAD

**Issue by ViktorHofer** [dotnet/runtime#58343]
Similar to what was being discussed in https://github.com/dotnet/runtime/issues/54964.

See https://github.com/dotnet/runtime/blob/9703660baa08914773b26e413e361c8ce04e6d94/src/libraries/Directory.Build.targets#L89-L95

The ILLinkTrimAssembly property is conditioned on NetCoreAppCurrent inbox assets only. Some packages contain both inbox and OOB assets. The illink optimization will be lost for net6.0 when NetCoreAppCurrent means net7.0 and new packages will be released.

@joperezr @stephentoub why do we only trim assemblies inside the shared framework?

**Comment by stephentoub** [comment]
> why do we only trim assemblies inside the shared framework?

cc: @eerhardt

**Comment by eerhardt** [comment]
> why do we only trim assemblies inside the shared framework?

It hasn't changed since it was first integrated into our build. See https://github.com/dotnet/corefx/pull/17825#commitcomment-29425831.

Other reasons I can imagine:

1. Build time tradeoff - According to https://github.com/dotnet/corefx/pull/17825#issuecomment-291566915, at the time the initial integration increased the local build time by 35%. We build a lot more libraries now - i.e. Microsoft.Extensions.*
2. If we think it is valuable for our OOB libraries to get trimmed before they are packaged, there are reasons to believe that libraries outside of dotnet/runtime would benefit as well. So at that point, it shouldn't be a dotnet/runtime build infrastructure feature, but instead a .NET SDK feature. That way it is fully supported by the tooling.
3. Do we think the benefits are going to out weigh the risk/cost? As you go up the stack, the amount of statically trimmable code decreases. If we have libraries with lots of "dead code" - we should just remove that dead code. At the shared framework (and also runtimepack) level, we can even trim the assemblies for OS-arch specific configurations as well. (However, note that idea has been rejected so far - see https://github.com/dotnet/runtime/issues/31712.).

@vitek-karas @marek-safar - any thoughts on this topic?

**Comment by ViktorHofer** [comment]
> We currently don't run any tests for "net6.0 when NetCoreAppCurrent means net7.0". So it is rather dangerous to be trimming these assemblies.

Agreed, I'm very much concerned that we don't test non current .NETCoreApp tfms, which is why filed https://github.com/dotnet/runtime/issues/54639. That said, isn't it even more dangerous to create an artificial difference between `$(NetCoreAppCurrent)` (linked) and previous .NETCoreApp assemblies (unlinked) as we only test `$(NetCoreAppCurrent)` which is linked?

Also, the moment we ship new packages (i.e. NET6), anyone who isn't on the very latest shared framework (including servicing releases) will suffer from performance regressions based on consuming unlinked assemblies. Are we fine with that?

**Comment by ViktorHofer** [comment]
Note that the actual set of libraries which are inbox and ship as part of packages is quite small:
- System.Collections.Immutable
- System.Diagnostics.DiagnosticSource
- System.Formats.Asn1
- System.Net.Http.Json
- System.Reflection.Metadata
- System.Runtime.CompilerServices.Unsafe
- System.Text.Encoding.CodePages
- System.Text.Encodings.Web
- System.Text.Json
- System.Threading.Channels
- System.Threading.Tasks.Dataflow

I think we should at least trim all .NETCoreApp inbox assemblies. At the moment, only the "best applicable" asset is trimmed. As we produce our packages on a Windows leg, inbox libraries which are RID specific are only trimmed when targeting Windows. This means that even for `$(NetCoreAppCurrent)` assemblies, the linker optimization might be lost if the assembly targets a non Windows RID.

Let's fix https://github.com/dotnet/runtime/issues/54639 and add non-current .NETCoreApp test configurations for the libraries in the list above. We should do that in the main branch and then back-port into release/6.0.

**Comment by eerhardt** [comment]
>  isn't it even more dangerous to create an artificial difference between $(NetCoreAppCurrent) (linked) and previous .NETCoreApp assemblies (unlinked) as we only test $(NetCoreAppCurrent) which is linked?

I wouldn't be able to say which is "more dangerous" without testing. Since we don't use the linker at all when in "library mode" against previous .NETCoreApp assemblies, I wouldn't put it as "safer" than any other scenario. Another thought is that leaving IL in an assembly typically works more often than taking IL out.

> Also, the moment we ship new packages (i.e. NET6), anyone who isn't on the very latest shared framework (including servicing releases) will suffer from performance regressions based on consuming unlinked assemblies. Are we fine with that?

What kind of performance regressions are you referring to? In the above list, how many of those assemblies have enough dead code in them to actually cause any measurable regression?

On a separate note - in the scenario you list, those assemblies are no longer crossgen'd when they come from NuGet packages, which will cause an even greater performance regression than having a few KBs of dead IL.

**Comment by joperezr** [comment]
> I think we should at least trim all .NETCoreApp inbox assemblies. At the moment, only the "best applicable" $(NetCoreAppCurrent) asset is trimmed.

I agree with @ViktorHofer that with the library set above, we should just turn on trim to all netcoreapp configs (not just the best applicable). There is no testing concern there as we have our unit test coverage testing against the trimmed version of the library, so shipping an un-trimmed version as oob (not tested) that can replace the trimmed one (tested) doesn't sound great. We should consider what to do with regards to other TFMs and other libraries, and check perf impact, but I think that at least for .NET 6, it would be good to enable trimming for all netcoreappcurrent configs for the set that @ViktorHofer showed above.

**Comment by eerhardt** [comment]
Note, as @vitek-karas suggested offline, an alternative approach would be to not trim those libraries at all (the libraries that ship both inbox and OOB), even "the best applicable". If the concern is to be consistent, not trimming at all would make it consistent.

**Comment by joperezr** [comment]
While that is true, if we are thinking about what do we want to ship in .NET 6, I'd be more comfortable with the approach of trimming always for those as opposed to not trimming. Reason being that we've been testing these assemblies trimmed for the whole product cycle, so seems a bit risky to change that this late.

**Comment by vitek-karas** [comment]
I'm not sure what's best for 6 - definitely the least changes. How did we ship in 5 - we had trimming back then as well... if we shipped in this "semi-trimmed" state in 5, I would probably leave it be.

I would definitely agree with @eerhardt that not-trimming is safer than trimming.

As for 7+, as @eerhardt already noted, I think we should reconsider trimming framework assemblies. There's value in it for some of them, but the value is very likely lower the higher in the stack the library goes. In the case of OOB packages, I think we just should not trim them... at all. So the assemblies which ship in them should never be trimmed for consistency and simplicity. We should obviously measure the size impact, but I doubt it's going to be important enough. (For the fun of it, I just measured the diff on debug build Windows x64, size saving for framework assemblies 64KB total, for OOB assemblies 0).

Performance wise - the size doesn't matter (at least not if we're talking about couple of KBs). The R2R is a much bigger impact.

As for sharing this tech with other assemblies/repos:
- The "analysis" run of the trimmer, we should share that eventually - first via Arcade and maybe one day even make it a public feature of the SDK.
- The "trim" run of the trimmer - I don't think it's worth the trouble. I'm not even sure it makes much of a different for the runtime repo, I doubt it will have any impact on higher-level repos.

**Comment by ViktorHofer** [comment]
Opened https://github.com/dotnet/runtime/issues/58500 to discuss the R2R regression when consuming package assets.

**Comment by joperezr** [comment]
FWIW, I checked the above set of assemblies and seems like the libraries with potential issues where we could ship a not-linked assembly but only tested a linked one, are:
- System.Security.AccessControl
- System.Text.Enconding.CodePages : We ship net6.0 asset which isn't linked as we link the net6.0-windows config instead
- System.Text.Encondings.Web: We ship the net6.0-Browser asset in a package which isn't linked, as the linked asset for the build would be net6.0

The most important one above is the last one, as the other ones are just PNSE assemblies which we aren't linking. For the last one, we should ensure that the Browser asset get's linked on the Windows build too in order to make sure that we are testing what we are shipping.

**Comment by ericstj** [comment]
In that case what it might mean is "app size increases when consuming System.Text.Encodings.Web package on browser" right?  Maybe it's not super critical since browser will get linked anyways, though with different settings.

**Comment by joperezr** [comment]
True, I guess the uneasy feeling for that one is that we are only testing the linked asset, but shipping an un-linked that will be used if the package is consumed. My suggestion of fixing it and marking as the "important one" is mostly due to the fact that we are shipping assets we don't test in that case.

**Comment by ericstj** [comment]
Let's make a call on this for 6.0.0 or move to 7.0.0.

**Comment by ViktorHofer** [comment]
1. For 6.0 I would like us to enable trimming NetCoreAppCurrent inbox source assemblies but exclude PNSE assemblies for: https://github.com/dotnet/runtime/pull/58345 (but change the `$(TargetFrameworkIdentifier)` condition to `$([MSBuild]::VersionGreaterThanOrEquals($(TargetFrameworkVersion), '$(NETCoreAppCurrentVersion)'))` to only trim NetCoreAppCurrent.
2. For 7.0 I think it makes sense to change this early to trim all inbox .NETCoreApp source assets assemblies: https://github.com/dotnet/runtime/pull/58345.

I'm fine with doing the change (nr 2 is already in PR and just needs to be resurrected).

**Comment by ericstj** [comment]
@ViktorHofer are you still planning to do something for 6.0 GA?
